### ISSUE TITLE
Add Bun.eval

### ISSUE THREAD

**Issue by FreePhoenix888** [oven-sh/bun#5105]
### What is the problem this feature would solve?

We would be able to use run code:
```ts
const functionExpression = Bun.eval(`() => {
  const result: number = 1+1;
  return result
}`)
const result = functionExpression()
```

### What is the feature you are proposing to solve the problem?

Add Bun.eval

**Comment by v3nntrnx** [comment]
Neat idea, an eval with typescript support would be really nice

**Comment by FreePhoenix888** [comment]
> Neat idea, an eval with typescript support would be really nice

+++. Our team desperately need it

**Comment by jhmaster2000** [comment]
This idea has been discussed a few times in the past, it will probably be done eventually but not very high priority right now.

In the meantime at least there is `Bun.Transpiler` you can use as a middleman between your TS and the vanilla `eval` (this might be what you already do as I recall you mentioning transpiling but still putting it out there as reference for others). This is also more-or-less what the REPL does too even.

**Comment by FreePhoenix888** [comment]
> This idea has been discussed a few times in the past, it will probably be done eventually but not very high priority right now.
> 
> In the meantime at least there is `Bun.Transpiler` you can use as a middleman between your TS and the vanilla `eval` (this might be what you already do as I recall you mentioning transpiling but still putting it out there as reference for others). This is also more-or-less what the REPL does too even.

Is there a reason to use `Bun.Transpiler` instead of `transpileModule` from `typescript`? @jhmaster2000

**Comment by FreePhoenix888** [comment]
The main my problem in this solution is that I do not know how to handle source maps to show actual lines in code instead of wrong lines of code caused of transpilation

**Comment by v3nntrnx** [comment]
> Is there a reason to use Bun.Transpiler instead of transpileModule from typescript? 

Yes, Bun.Transpiler is bun's transpiler which is *blazingly* faster than tsc (typescript package)

**Comment by jhmaster2000** [comment]
> > This idea has been discussed a few times in the past, it will probably be done eventually but not very high priority right now.
> > In the meantime at least there is `Bun.Transpiler` you can use as a middleman between your TS and the vanilla `eval` (this might be what you already do as I recall you mentioning transpiling but still putting it out there as reference for others). This is also more-or-less what the REPL does too even.
> 
> Is there a reason to use `Bun.Transpiler` instead of `transpileModule` from `typescript`? @jhmaster2000

> The main my problem in this solution is that I do not know how to handle source maps to show actual lines in code instead of wrong lines of code caused of transpilation

Well it is way faster and native so you don't need an external runtime dependency on the entirety of `typescript` (very bulky) but source maps is a good point, afaik Bun.Transpiler is indeed lacking a little on those still, so I guess it would be better for you to hold off on it for a while longer still, but it'll definitely be sorted out sometime.

**Comment by xhyrom** [comment]
I don'ŧ see any good reason why to implement another eval function. If it's because typescript, you could use `Bun.Transpiler`, but you can't, because it eliminates dead code and this option can't be disabled. Adding something like `deadCodeElimination` option into `Bun.Transpiler` sounds like a better solution. Using eval is a bad approach.

This would be possible after the `deadCodeElimination` option:
```ts
function customEval(code: string) {
  const transpiler = new Bun.Transpiler({
    loader: "ts",
    deadCodeElimination: false,
  });
  const transpiled = transpiler.transformSync(code);

  return eval(transpiled);
}

const f = customEval("() => { return 5; }");
f();
```

**Comment by LowByteFox** [comment]
A temporary solution could be 
```ts
Bun.eval = (code: string) => {
    const transpiler = new Bun.Transpiler({
        loader: "ts"
    });

    return eval(transpiler.transformSync(`eval((${code}))`));
}

const functionExpression = Bun.eval(`() => {
  const result: number = 1+1;
  return result;
}`)

const result = functionExpression();
```
You can take the `Bun.Transpiler` out of the function just so you have only one instance of it

**Comment by jhmaster2000** [comment]
> I don'ŧ see any good reason why to implement another eval function. If it's because typescript, you could use `Bun.Transpiler`, but you can't, because it eliminates dead code and this option can't be disabled. Adding something like `deadCodeElimination` option into `Bun.Transpiler` sounds like a better solution. Using eval is a bad approach.
> 
> This would be possible after the `deadCodeElimination` option:
> 
> ```ts
> function customEval(code: string) {
>   const transpiler = new Bun.Transpiler({
>     loader: "ts",
>     deadCodeElimination: false,
>   });
>   const transpiled = transpiler.transformSync(code);
> 
>   return eval(transpiled);
> }
> 
> const f = customEval("() => { return 5; }");
> f();
> ```

It's true that the dead code elimination toggle for Transpiler is very needed, but it doesn't solve another issue which was what the new eval function was originally talked about for, which is eval with top level await:
```js
eval("await 5"); // fail, and no amount of transpilation can fix this
asyncEval("await 5"); // this was the original proposed name
```
Direct TS support would be a bonus.

**Comment by FreePhoenix888** [comment]
> Using eval is a bad approach.

I know about problems related to `eval` but our team has a reason to do it. We are [deep-foundation](https://github.com/deep-foundation) and we bring a lot of languages together in our associative system. We have a set of docker-isolation providers which get some code and call that with some context. [Here is an example](https://github.com/deep-foundation/js-docker-isolation-provider). Right now we have some problems with typescript cause we have to handle source-maps ourselves after transpiling ts to js and evaluating it and this is the reason why I wanted to know whether Bun is going to create Bun.eval (I guess if it will be created it will support correct source-mapping)

**Comment by FreePhoenix888** [comment]
> > Is there a reason to use Bun.Transpiler instead of transpileModule from typescript?
> 
> Yes, Bun.Transpiler is bun's transpiler which is _blazingly_ faster than tsc (typescript package)



> > > This idea has been discussed a few times in the past, it will probably be done eventually but not very high priority right now.
> > > In the meantime at least there is `Bun.Transpiler` you can use as a middleman between your TS and the vanilla `eval` (this might be what you already do as I recall you mentioning transpiling but still putting it out there as reference for others). This is also more-or-less what the REPL does too even.
> > 
> > 
> > Is there a reason to use `Bun.Transpiler` instead of `transpileModule` from `typescript`? @jhmaster2000
> 
> > The main my problem in this solution is that I do not know how to handle source maps to show actual lines in code instead of wrong lines of code caused of transpilation
> 
> Well it is way faster and native so you don't need an external runtime dependency on the entirety of `typescript` (very bulky) but source maps is a good point, afaik Bun.Transpiler is indeed lacking a little on those still, so I guess it would be better for you to hold off on it for a while longer still, but it'll definitely be sorted out sometime.

@jhmaster2000 @TiranexDev as I understand even if I transpile ts code to js by using bun I still should have typescript installed to see typescript errors in my IDE?

**Comment by FreePhoenix888** [comment]
> A temporary solution could be
> 
> ```ts
> Bun.eval = (code: string) => {
>     const transpiler = new Bun.Transpiler({
>         loader: "ts"
>     });
> 
>     return eval(transpiler.transformSync(`eval((${code}))`));
> }
> 
> const functionExpression = Bun.eval(`() => {
>   const result: number = 1+1;
>   return result;
> }`)
> 
> const result = functionExpression();
> ```
> 
> You can take the `Bun.Transpiler` out of the function just so you have only one instance of it

I like it. Is there a possibility to add source maps here to have actual code-line instead of having lines of transpiled  code? @LowByteFox

**Comment by Kleywalker** [comment]
The workaround does not work in common cases. Like [here](https://github.com/Kleywalker/bun-eval).
Also
"index.ts is an ES module, and exports is not a variable defined.

eval doesnt inherit scope either" [#7635](https://github.com/oven-sh/bun/issues/7635)

and

"--target=node doesnt create a CommonJS module. eval does not support ES module syntax." [#7636](https://github.com/oven-sh/bun/issues/7636)

**Comment by paperclover** [comment]
fwiw, i am against adding this function in any way to Bun, `eval` is a bad pattern that should be discouraged. `Bun.Transpiler` can be used in combination with `eval`

**Comment by Kleywalker** [comment]
I would like to eval simple code lines like `import Case from 'case';` and `export default 42;` using bun. What do I have to do?
